_topics/en/op_checksigfromstack.md

```diff
@@ -0,0 +1,107 @@
+---
+title: OP_CHECKSIGFROMSTACK
+
+## Optional.  An entry will be added to the topics index for each alias
+#aliases:
+#  - Foo
+
+## Required.  At least one category to which this topic belongs.  See
+## schema for options
+categories:
+  - Scripts and Addresses
+  - Soft Forks
+
+## Required.  Use Markdown formatting.  Only one paragraph.  No links allowed.
+## Should be less than 500 characters
+excerpt: >
+  **OP_CHECKSIGFROMSTACK (OP_CSFS)** is an opcode on
+  ElementsProject.org-based sidechains that is sometimes proposed for
+  implementation on Bitcoin.  The opcode allows checking whether a
+  signature signs an arbitrary message.  The opcode takes three
+  parameters: a signature, a message, and a public key.
+
+## Optional.  Use Markdown formatting.  Multiple paragraphs.  Links allowed.
+extended_summary: |
+  In Bitcoin's existing signature-checking opcodes (such as
+  `OP_CHECKSIG`), the message is required to be certain parts of the the
+  serialized spending transaction plus some additional verification
+  data.  That allows proving that a signature not only corresponds to a
+  certain pubkey but that the private key used to generate both of those
+  objects was used to authorize the spend.
+
+  If the same pubkey and signature pair are valid both with `OP_CSFS`
```

Comment:
I found this description quite confusing on first read. It wasn't initially clear to me that this paragraph and the rest of the summary were about a scheme to use OP_CSFS to do transaction introspection.

Is it possible to use subheadings in the summary? If so, perhaps this would be clearer as follows:

```
<generic description of OP_CSFS and existing OP_CHECKSIG opcodes>

#### Using OP_CSFS for transaction introspection

One of the most interesting applications of OP_CSFS is to be able to _introspect_ arbitrary parts of the transaction.

<describe what introspection means, using CLTV as an example>

<describe how OP_CSFS can be used for transaction introspection>

<some examples of what tx introspection could allow>
```

Reply:
1. Subheadings can be used in a summary, we already use one on the codesep topic: https://bitcoinops.org/en/topics/op_codeseparator/#later-changes-and-proposals

2. I don't mind adding the subhead you suggest, but I think it suggests that there are multiple uses for CSFS---yet I'm unaware of any that can't be done more efficiently some other way (e.g. paying for signatures to arbitrary messages can be done using adapter signatures).

Reply:
Er, for point n2, I meant to ask: are there any other uses you're aware of?  If not, I think I'll just try rewriting to see if I can make the text more congruous.

Reply:
> I'm unaware of any that can't be done more efficiently some other way (e.g. paying for signatures to arbitrary messages can be done using adapter signatures).

Perhaps that's worth stating explicitly in the summary? I (not having looked into OP_CSFS) assumed that there would be other use cases for checking signatures over arbitrary messages in script.

Reply:
I asked about other uses in #sidechains-dev and Russell O'Connor replied, "Prior to introspection, I always assumed that CHECKSIGFROMSTACK was for oracle, as in a trusted third party that signs the outcome of various events in order to facilitate prediction / futures markets."

Although I think DLCs would be more private and efficient for that, there's an admirable simplicity to a contract that's `<hash('The agreement...')> <oracle_pubkey> OP_CSFS` (and DLCs weren't proposed when CSFS was first made available on Elements).

Anyway, I'll rewrite the topic and try to make the article clearer.